url433bec1e56_sarobe_DifficultyInvestigation_tgz-pJ8-fr_inria_optimization_cmaes_examples_CMAExample1J8/src/fr/inria/optimization/cmaes/CMAEvolutionStrategy.java

```diff
@@ -0,0 +1,2704 @@
+            double res[] = new double[v.length];
+            for(int i = 0; i < v.length; ++i)
+                res[i] = Math.abs(v[i]);
+            return res;
+        }
+    } // MyMath
+    
+    class Timing {
+        Timing(){
+            birth = System.currentTimeMillis();
+            start = birth; // on the save side 
+        }
+        long birth; // time at construction, not really in use
+        long start; // time at end of init()
+        long starteigen; // time after flgdiag was turned off, ie when calls to eigen() start
+        long eigendecomposition = 0; // spent time in eigendecomposition
+        long writedefaultfiles = 0;        // spent time in writeToDefaultFiles
+    }
+    Timing timings = new Timing();
+    
+    /* flgforce == 1 force independent of time measurments, 
+     * flgforce == 2 force independent of uptodate-status
+     */
+    void eigendecomposition(int flgforce) {
+        /* Update B and D, calculate eigendecomposition */
+        int i, j;
+        
+        if (countCupdatesSinceEigenupdate == 0 && flgforce < 2)
+        	return;
+
+    	//           20% is usually better in terms of running *time* (only on fast to evaluate functions)
+    	if (!flgdiag && flgforce <= 0 && 
+    			(timings.eigendecomposition > 1000 + options.maxTimeFractionForEigendecomposition 
+    					* (System.currentTimeMillis() - timings.starteigen)
+    					|| countCupdatesSinceEigenupdate < 1. / sp.getCcov() / N / 5.)) 
+    		return;
+
+    	if (flgdiag) {
+    		for (i = 0; i < N; ++i) {
+    			diagD[i] = Math.sqrt(C[i][i]);
+    		}
+    		countCupdatesSinceEigenupdate = 0;
+        	timings.starteigen = System.currentTimeMillis(); // reset starting time
+        	timings.eigendecomposition = 0;             // not really necessary
+    	} else {
+    		// set B <- C
+    		for (i = 0; i < N; ++i)
+    			for (j = 0; j <= i; ++j)
+    				B[i][j] = B[j][i] = C[i][j];
+
+    		// eigendecomposition
+    		double [] offdiag = new double[N];
+    		long firsttime = System.currentTimeMillis();
+    		tred2(N, B, diagD, offdiag);
+    		tql2(N, diagD, offdiag, B);
+    		timings.eigendecomposition += System.currentTimeMillis() - firsttime;
+
+    		if (options.checkEigenSystem > 0)
+    			checkEigenSystem( N,  C, diagD, B); // for debugging 
+
+    		// assign diagD to eigenvalue square roots
+    		for (i = 0; i < N; ++i) {
+    			if (diagD[i] < 0) // numerical problem?
+    				error("an eigenvalue has become negative");
+    			diagD[i] = Math.sqrt(diagD[i]);
+    		}
+    		countCupdatesSinceEigenupdate = 0;
+    	} // end Update B and D
+    	if (math.min(diagD) == 0) // error management is done elsewhere
+    		axisratio = Double.POSITIVE_INFINITY;
+    	else
+    		axisratio = math.max(diagD) / math.min(diagD);
+
+    } // eigendecomposition
+
+
+    /* ========================================================= */
+    int
+    checkEigenSystem( int N,  double C[][], double diag[], double Q[][]) 
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 16. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.